src/main/ruby/truffleruby/core/time.rb

```diff
@@ -255,6 +255,49 @@ def ceil(places = 0)
     time + (ceiled - original)
   end
 
+  DEFAULT_DECONSTRUCT_KEYS = %i[year month day yday wday hour min sec subsec dst zone]
+
+  def deconstruct_keys(array_of_names)
+    unless Primitive.nil?(array_of_names) || Primitive.is_a?(array_of_names, Array)
+      raise TypeError, "wrong argument type #{Primitive.class(array_of_names)} (expected Array or nil)"
+    end
+
+    if Primitive.nil?(array_of_names)
+      array_of_names = DEFAULT_DECONSTRUCT_KEYS
+    end
+
+    ret = {}
+
+    array_of_names.each do |key|
+      case key
+      when :year
+        ret[:year] = year
+      when :month
+        ret[:month] = month
```

Comment:
The `ret[:year] =` etc could be moved out of the case like `ret[key] = case key`.
This avoids the duplication and significantly reduces the memory footprint and # of AST nodes for this method.
@andrykonchin Could you do it (since we have little time before the feature freeze)?

Reply:
I am on it 👍

Reply:
Ah, that's a good catch!